pkg/runtime/serializer/versioning/versioning.go

```diff
@@ -225,51 +137,95 @@ func (c *codec) Decode(data []byte, defaultGVK *unversioned.GroupVersionKind, in
 // Encode ensures the provided object is output in the appropriate group and version, invoking
 // conversion if necessary. Unversioned objects (according to the ObjectTyper) are output as is.
 func (c *codec) Encode(obj runtime.Object, w io.Writer) error {
-	if _, ok := obj.(*runtime.Unknown); ok {
+	switch t := obj.(type) {
+	case *runtime.Unknown:
+		if gv, ok := runtime.PreferredGroupVersion(c.encodeVersion); ok {
+			t.APIVersion = gv.String()
+		}
+		return c.encoder.Encode(obj, w)
+	case *runtime.Unstructured:
+		if gv, ok := runtime.PreferredGroupVersion(c.encodeVersion); ok {
+			t.SetAPIVersion(gv.String())
```

Comment:
Same question-- it seems like the destination version should depend on the group versioner and the object's current, in-memory gvk. (I assume that the below Encode call uses this api version as a hint as to what it should convert the object to, or if not this makes no sense to me)